File: problems/05-stacks-queues/easy/lc-020-valid-parentheses/solution.cpp
#include <bits/stdc++.h>

class Solution {
public:
    bool isValid(string s) {
        std::stack<char> brackets;
        std::string::iterator it;

        for (it = s.begin(); it != s.end(); ++it) {
            char c = *it;
            if (*it == '(' || *it == '[' || *it == '{') {
                brackets.push(*it);
            }
            else {
                if ( brackets.empty() ) return false;

                char current = brackets.top();
                brackets.pop();

                if ( current == '{' && c != '}' ) return false;
                if ( current == '[' && c != ']' ) return false;
                if ( current == '(' && c != ')' ) return false;
            }
        }
        return brackets.empty();
    }   
};

File: problems/05-stacks-queues/easy/lc-020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        match = {')': '(', ']': '[', '}': '{'}

        for c in s:
            if c in "([{":
                st.append(c)
            else:
                if not st or st[-1] != match.get(c):
                    return False
                st.pop()

        return not st
